lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,881 @@
+			})
+		))
+	}
+}
+
+impl IHandshakeState for InitiatorAwaitingActTwoState {
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (receiver)
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-three (sender)
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		let mut act_two_builder = self.act_two_builder;
+		let remaining = act_two_builder.fill(input);
+
+		// Any payload larger than ACT_ONE_TWO_LENGTH indicates a bad peer since responder data
+		// is required to generate post-authentication messages (so it can't come before we transition)
+		if remaining.len() != 0 {
+			return Err("Act Two too large".to_string());
+		}
+
+		// In the event of a partial fill, stay in the same state and wait for more data
+		if !act_two_builder.is_finished() {
+			assert_eq!(remaining.len(), 0);
+			return Ok((
+				None,
+				HandshakeState::InitiatorAwaitingActTwo(Self {
+					initiator_static_private_key: self.initiator_static_private_key,
+					initiator_static_public_key: self.initiator_static_public_key,
+					initiator_ephemeral_private_key: self.initiator_ephemeral_private_key,
+					responder_static_public_key: self.responder_static_public_key,
+					chaining_key: self.chaining_key,
+					hash: self.hash,
+					act_two_builder
+				})
+			));
+		}
+
+		let initiator_static_private_key = self.initiator_static_private_key;
+		let initiator_static_public_key = self.initiator_static_public_key;
+		let initiator_ephemeral_private_key = self.initiator_ephemeral_private_key;
+		let responder_static_public_key = self.responder_static_public_key;
+		let hash = self.hash;
+		let chaining_key = self.chaining_key;
+		let act_two = Act::from(act_two_builder);
+
+		let (responder_ephemeral_public_key, hash, chaining_key, temporary_key) = process_act_message(
+			&act_two,
+			&initiator_ephemeral_private_key,
+			chaining_key,
+			hash,
+		)?;
+
+		let mut act_three = EMPTY_ACT_THREE;
+
+		// start serializing act three
+		// 1. c = encryptWithAD(temp_k2, 1, h, s.pub.serializeCompressed())
+		chacha::encrypt(&temporary_key, 1, &hash, &initiator_static_public_key.serialize(), &mut act_three[1..50]);
+
+		// 2. h = SHA-256(h || c)
+		let hash = concat_then_sha256!(hash, act_three[1..50]);
+
+		// 3. se = ECDH(s.priv, re)
+		let ecdh = ecdh(&initiator_static_private_key, &responder_ephemeral_public_key);
+
+		// 4. ck, temp_k3 = HKDF(ck, se)
+		let (chaining_key, temporary_key) = hkdf::derive(&chaining_key, &ecdh);
+
+		// 5. t = encryptWithAD(temp_k3, 0, h, zero)
+		chacha::encrypt(&temporary_key, 0, &hash, &[0; 0], &mut act_three[50..]);
+
+		// 6. sk, rk = HKDF(ck, zero)
+		let (sending_key, receiving_key) = hkdf::derive(&chaining_key, &[0; 0]);
+
+		// 7. rn = 0, sn = 0
+		// - done by Conduit
+		let conduit = Conduit::new(sending_key, receiving_key, chaining_key);
+
+		Ok((
+			Some(Act::Three(act_three)),
```

Comment:
You can add `Send m = 0 || c || t`

Reply:
Resolved in `review: Non-controversial renames, nits, small fixups`